File: Stage06-FromFile/Lua/player.lua
local Player = {}
local Shared = require("shared")
--static player class - no constructor required
Player.Name 		= ""
Player.Health 		= 100
Player.Strength 	= 100
Player.Character 	= ""
Player.Characters 	= {'fighter','wizard','ninja','priest'}
Player.Inventory 	= {}
Player.ItemInHand 	= {}

local function TableIndex(tbl, item)
	for key, value in pairs(tbl) do
		if item == value then
			return key			-- found match
		end
	end
	
	return 0
end

function Player.AddToInventory(item)
	if TableIndex(Player.Inventory, item) == 0 then
		table.insert(Player.Inventory, item)
	end
end

function Player.Attack(here)
	local enemy = here:GetEnemy()
	local message = "You attack the "..enemy
	--[[
	if item_in_hand != "":
		if isinstance(shared.items[item_in_hand], weapon.Weapon):
			damage = shared.items[item_in_hand].damage
			message += f"with the {item_in_hand} inflicting {damage} damage points"
	else:
		message += " with your fist inflicting 10 damage points"
	]]
	local damage = 10
	local useItem = "your fist"
	for _, item in ipairs(Player.Inventory) do
		if Shared.Items[item].type == "Weapon" then
			damage = Shared.Items[item]:GetDamage()
			useItem = " the ".. item
			break
		end
	end
	
	message = message .. " with "..useItem.." inflicting "..damage.." damage points"
			
	Shared.Enemies[enemy]:ReceiveAttack(damage)
	
	return message	
end

function Player.DisplayInventory()
	--[[ display player;s inventory ]]
	if #Player.Inventory == 0 then
		print("Your inventory is empty")
	else
		print("In your inventory you have:")
		print(table.concat(Player.Inventory, ", "))
	end
end

function Player.DisplayPlayer()
	--[[ main use for debug. prints all player properties ]]
	print(string.rep('═', 80))
	print("Player properties:")
	print(string.rep('═', 80))
	print("Characters available: ".. table.concat(Player.Characters, ", "))
	print(string.rep('═', 80))
	print("Name:                 "..Player.Name)
	print("Health:               "..Player.Health)
	print("Strength:             "..Player.Strength)
	print("Character:            "..Player.Character)
	print("Inventory:            "..table.concat(Player.Inventory, ", "))
	print(string.rep('═', 80))
end

function Player.GetProperty(propertyName)
	if propertyName:find("name") ~= nil then
		return Player.Name
	elseif propertyName:find("health") ~= nil then
		return Player.Health
	elseif propertyName:find("strength") ~= nil then
		return Player.Strength
	elseif propertyName:find("character") ~= nil then
		return Player.Character
	end
end

function Player.ReceiveAttack(damage)
	Player.Health = Player.Health - damage
	if Player.Health <= 0  then
		Player.Health = 0
	end
end

function Player.RemoveFromInventory(item)
	local index = TableIndex(Player.Inventory, item)
	if index > 0 then
		table.remove(Player.Inventory, index)
	end
end

function Player.UpdateStats(characterIndex)
	--[[ modify health and strength depending on character selected ]]
	Player.Health = Player.Health +  characterIndex * 2
	Player.Strength = Player.Health - characterIndex * 2
end

return Player

File: Stage07-Improvements/C#/Player.cs
using System;
using System.Collections.Generic;

namespace Adventure_06_Improvements
{
    internal static class Player
    {
        #region Static Properties
        public static string Name = "";
        public static int Health = 100;
        public static int Strength = 100;
        public static string Character = "";
        public static string ItemInHand = "";
        public static List<string> Characters = new List<string>();
        public static List<string> Inventory = new List<string>();
        #endregion
        #region Methods
        public static void AddToInventory(string item)
        {
            /// add an item to player inventory ///
            if (!Inventory.Contains(item))
                Inventory.Add(item);
        }
        public static string Attack(Location here, string item)
        {
            /// Item already chosen, use here to attack enemy ///
            string enemy = here.Enemy;
            string message = $"You attack the {enemy}";
            int damage = 5;

            if(Shared.Items[item] is Weapon)                // Is the item a weapon?
            {
                Weapon weapon = (Weapon)Shared.Items[item]; // cast Item to Weapon to obtain Damage
                damage = weapon.Damage;
                item = $" the {item}";
            }
            message += $" with {item} inflicting {damage} damage points";
            Shared.Enemies[enemy].ReceiveAttack(damage);

            return message;
        }
        public static string GetProperty(string propertyName)
        {
            ///  used by narrator class to get player properties ///
            if (propertyName.Contains("name"))
                return Name;
            else if (propertyName.Contains("health"))
                return Health.ToString();
            else if (propertyName.Contains("strength"))
                return Strength.ToString();
            else if (propertyName.Contains("character"))
                return Character;
            return "";
        }
        public static void ReceiveAttack(int damage)
        {
            /// reduce health to min of 0 ///
            Health -= damage;
            if(Health < 0)
                Health = 0;
        }

        public static void RemoveFromInventory(string item)
        {
            ///  remove an item from player inventory ///
            if(Player.Inventory.Contains(item))
                Player.Inventory.Remove(item);
        }
        public static void UpdateStats(int characterIndex)
        {
            ///  modify health and strength depending on character selected ///
            Health += characterIndex * 2;
            Strength += characterIndex * 2;
        }
        #endregion
    }
}
